R/modify.R

```diff
@@ -93,41 +93,18 @@ modify <- function(.x, .f, ...) {
 modify.default <- function(.x, .f, ...) {
   .f <- as_mapper(.f, ...)
 
-  for (i in seq_along(.x)) {
-    list_slice2(.x, i) <- .f(.x[[i]], ...)
+  if (vec_is_list(.x) || is.data.frame(.x)) {
+    out <- map(vec_proxy(.x), .f, ...)
+    vec_restore(out, .x)
+  } else if (vec_is(.x)) {
+    map_vec(.x, .f, ..., .ptype = .x)
+  } else if (is.null(.x) || is.list(.x)) {
+    .x[] <- map(.x, .f, ...)
+    .x
+  } else {
+    cli::cli_abort("Don't know how to modify {.obj_type_friendly {.x}}.")
```

Comment:
Can we use a "can't" form here?

```
! Can't modify a function.
```

Actually a "must" form works well I think:

```
! `.x` must be a vector, not a function.
```

Reply:
Do you think I should just ignore the fact it's a generic? It is unlikely that people have provided methods. Do you think I verify that there aren't methods on CRAN then remove the genericity?

Reply:
There are five packages with `S3method(modify,` in their namespace (https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Acran+path%3A%2F%5ENAMESPACE%2F+%2FS3method%5C%28modify%2C%2F)

* heemod — own modify generic
* timbr — purrr generic
* purrr — duh
* tfarima — own generic
* yamlet — own generic

timbr's `modify.forest` looks considerably more complex than our implementations, so it's probably ok for it to be its own function: https://github.com/UchidaMizuki/timbr/blob/main/R/purrr.R